### ISSUE TITLE
Request: Provide full featured builds

### ISSUE THREAD

**Issue by schlomo** [ipxe/ipxe#175]
First of, I am really happy to see that iPXE is still going on!

Would it be possible to provide full featured builds at http://boot.ipxe.org/ instead of the default builds? Or in addition to the default builds?

I see two benefits for that:
1. easy place for users to download an iPXE version that can do (almost) everything that is there, helping to show case iPXE abilities and reducing the effort for users who don't need further customization
2. demonstrate which features can be combined in a build without troubles

I got to this point when setting up a universal network boot server for myself. I tried to find precompiled iPXE binaries that would be fairly recent and support HTTPS booting and framebuffer console. No luck.

I know that I can build my own but that is an extra step while I would like to provide easy to follow instructions for setting up such a universal boot server.

Since you in any case have an automated setup for building iPXE and uploading it to http://boot.ipxe.org/, I am asking you to add there also a "big" version of iPXE.

**Comment by MaxPeal** [comment]
+1 for a "big" version of iPXE with https hosted on http://boot.ipxe.org/big or so

**Comment by NiKiZe** [comment]
You have to define "full" because there is features that can't be enabled at the same time.

I agree that ipxe+Https should be great to have and maybe also an ipxe+https+IPv6 (mostly true for pcbios since it's not enabled by default)

**Comment by schlomo** [comment]
What would be an "as much as possible" build? Maybe without some exotic options? I am sure that if you let us know about the mutually exclusive features then we can come to a meaningful (= useful for many users) compromise. Just ask.

For example, features that require a matching setup on other systems probably won't be so important in a "big build". Or features that are closely related to seldom or special hardware.

**Comment by berlin4apk** [comment]
@NiKiZe  @mcb30 
Can you give us a hint to find all mutually exclusive features?

**Comment by NiKiZe** [comment]
I'd say no, don't include more than you need.
Sure a "bigger" version with often used extra features, such as vlan and framebuffer, but we need that information.

**Comment by NiKiZe** [comment]
Instead of arguing about a "big" version, how about that we at least start with a "more" build, where we request features that we know is useful?
* [https](https://ipxe.org/buildcfg/download_proto_https)
* [framebuffer](https://ipxe.org/buildcfg/console_framebuffer) along with console and png
* [param](https://ipxe.org/buildcfg/param_cmd)
* [vram](https://ipxe.org/cfg/vram) (pcbios only)

what else?

**Comment by schlomo** [comment]
Yes, this is a great idea!

**Comment by PicoMitchell** [comment]
I find these very useful:

For UEFI & BIOS:
poweroff Command: https://ipxe.org/buildcfg/poweroff_cmd
ping Command: https://ipxe.org/buildcfg/ping_cmd
cpuvendor and cpumodel Settings: https://github.com/ipxe/ipxe/issues/419 (undocumented)

And for BIOS only (since it doesn't work in UEFI last I checked):
memsize Setting: https://ipxe.org/buildcfg/memmap_settings

**Comment by berlin4apk** [comment]
I find this useful, together with the already mentioned, options:

NSLOOKUP_CMD
PCI_CMD
REBOOT_CMD
TIME_CMD
VLAN_CMD
IMAGE_TRUST_CMD
IMAGE_ARCHIVE_CMD
DOWNLOAD_PROTO_FILE
NVO_CMD
DIGEST_CMD
IMAGE_MEM_CMD

**Comment by NiKiZe** [comment]
iPXE 2.0.0 is released and the config structure has changed, several of the requested features are now enabled by default in efi builds (Will focus on these below), but not in pcbios builds due to size constraints.
I'll come back and update this post, so this is only a rough update on current state of enabled vs requested.
How many is still in need for pcbios builds?, and maybe a quick rationale of why is good.

```c
/* Disable protocols not historically included in BIOS builds */
#if defined ( PLATFORM_pcbios )
#undef DOWNLOAD_PROTO_HTTPS
#undef HTTP_AUTH_NTLM
#endif

/* Disable commands not historically included in BIOS builds */
#if defined ( PLATFORM_pcbios )
#undef CERT_CMD
#undef CONSOLE_CMD
#undef DIGEST_CMD
#undef NTP_CMD
#undef PARAM_CMD
#undef PCI_CMD
#undef USB_CMD
#undef VLAN_CMD
#endif
```

* ~~[https](https://ipxe.org/buildcfg/download_proto_https)~~
* ~~[framebuffer](https://ipxe.org/buildcfg/console_framebuffer) along with console and png~~
* ~~[param](https://ipxe.org/buildcfg/param_cmd)~~
* [vram](https://ipxe.org/cfg/vram) (pcbios only)
* [poweroff](https://ipxe.org/buildcfg/poweroff_cmd)
* [ping](https://ipxe.org/buildcfg/ping_cmd)
* ~~[cpuid](https://ipxe.org/buildcfg/cpuid_settings)~~ You can also get these thru smbios
* [memmap_settings](https://ipxe.org/buildcfg/memmap_settings)

NSLOOKUP_CMD
~~PCI_CMD~~
REBOOT_CMD
TIME_CMD
~~VLAN_CMD~~
IMAGE_TRUST_CMD
IMAGE_ARCHIVE_CMD
~~DOWNLOAD_PROTO_FILE default in EFI, not supported elsewhere~~
~~NVO_CMD~~
~~DIGEST_CMD~~
IMAGE_MEM_CMD

> If it's not a stopper for the UEFI Microsoft signing:
CONSOLE_VMWARE
VMWARE_SETTINGS

Use these for separate named configs builds instead.

**Comment by schlomo** [comment]
Thanks for keeping us posted, I'd like to argue again for a "as big as possible" build that provides easy access to as many features as possible for testing, experimentation and as a showcase.

If this works only over EFI then it is totally fine as nowadays EFI support is extremely widespread.

**Comment by mcb30** [comment]
> Thanks for keeping us posted, I'd like to argue again for a "as big as possible" build that provides easy access to as many features as possible for testing, experimentation and as a showcase.

I did put some time into considering the features that would be useful in a generally available build.  Which **specific** features are currently disabled that you think would generally useful for large numbers of users?

**Comment by schlomo** [comment]
Hi @mcb30 nice to hear from you! Please allow me to turn the question around:

As a new user exploring iPXE I'd be happy to get a binary download that is as little restricted as possible so that I don't need to think about it when playing around and discovering how iPXE can help me.

Therefore I'd strongly suggest to try creating a big build that has everything enabled that doesn't create a problem (e.g. for other modules) or that doesn't require site-specific additional configuration. That way we can expose more users to more iPXE features.

Maybe the thing to consider is the different perspective of seasoned iPXE users and developers vs. newbies: For you the question is "what's missing" and for newbies like me the questions is "what can I easily try out".

What is the harm in having a maximum build that brings as many features to the table as possible?

TBH, I haven't used iPXE in more than a cursory setting since 2020 when I was integrating it into our data center deployment automation and back then I ended up with an automated build customized for our required features. Back then a universal binary would have helped and I'm sure that also today a universal binary will help other newbies. Which is the context of me creating this issue back in 2020... That is why I personally can't tell you what specific feature is missing, beyond the feeling of missing out on potentially cool features without going down the road of a custom build.

Maybe for fun comparison: Most Linux distros ship "fully featured" kernel images so that nowadays nobody has to compile a kernel any more. I remember the times when compiling a custom kernel was standard procedure and TBH I don't miss them 😄

**Comment by NiKiZe** [comment]
Except for the technical aspects there is also a non zero cost in terms of questions. Now a certain level is needed to find and enable some features, which means that when something doesn't work there is a higher possibility of self help.
Let's say it doesn't work and to debug the issue we actually need to have the binary rebuilt, now we must not only help with the actual issue, but probably try to support in how to build it as well.

Except for this there is for example Wifi and NFS support that for security reasons shouldn't be enabled. https://github.com/ipxe/ipxe/pull/698#issuecomment-2031745651

So again, which features are anyone missing, and what is that feature used for?

**Comment by mcb30** [comment]
> Hi [@mcb30](https://github.com/mcb30) nice to hear from you! Please allow me to turn the question around:
> 
> As a new user exploring iPXE I'd be happy to get a binary download that is as little restricted as possible so that I don't need to think about it when playing around and discovering how iPXE can help me.

Yes, that's what I've aimed to do already with the default feature selection.  Anything that I expected people to want to use is enabled.  What's not enabled are niche use cases (such as the SLAM protocol that hasn't been meaningfully touched or tested since 2008), developer-only features (such as the `lotest` and `profstat` commands), and code that is not likely to work without substantial expertise (such as the long-outdated COMBOOT support).

> What is the harm in having a maximum build that brings as many features to the table as possible?

Excessive code size, and the fact that it would drag in lots of essentially old and unsupported code that is likely to malfunction.

**Comment by schlomo** [comment]
@mcb30 that sounds great and maybe solves already the main reason for me opening this ticket back then...

@NiKiZe I'd say that for experimentation Wifi and NFS are great features to include, better add a banner to that iPXE build pointing out that it is a demo build not intended for production use and that is has a lot of dangerous features enabled.
